Issue63 IntersectBy and ExceptBy overloads v2

JamesFaix opened the pull request:
Created IntersectBy operator.
Created similar IntersectKeys and ExceptKeys operators, which take a collection of keys as their second argument.
Created unit tests for IntersectBy, IntersectKeys, ExceptKeys, analogous to existing tests for ExceptBy

JamesFaix commented:
see also: [v1](https://github.com/morelinq/MoreLINQ/pull/163)

JamesFaix commented:
The last several commits address issues #124 , #125, and #80

fsateler commented on MoreLinq/ExceptAll.cs:
```diff
@@ -0,0 +1,186 @@
+        /// in the second sequence.
+        /// </summary>
+        /// <remarks>
+        /// This is a sequence operation; if multiple elements in <paramref name="first"/> are equal, 
```
I would add a note here that this is unlike `Except` from LINQ (I was wondering why we wanted `ExceptAll`...

fsateler commented on MoreLinq/ExceptAll.cs:
```diff
@@ -0,0 +1,186 @@
+        /// <param name="second">The sequence of keys which may prevent elements in
+        /// <paramref name="first"/> from being returned.</param>
+        /// <param name="keySelector">The mapping from source element to key.</param>
+        /// <param name="keyComparer">The key comparer. If <c>null</c>, uses the default TKey equality comparer.</param>
```
Missing <typeparamref> for annotating TKey.

fsateler commented on MoreLinq/ExceptAll.cs:
```diff
@@ -0,0 +1,186 @@
+        /// This is a sequence operation; if multiple elements in <paramref name="first"/> have
+        /// equal keys, all such elements are returned.
+        /// This operator uses deferred execution and streams the results, although
+        /// a set of keys from <paramref name="second"/> is immediately selected and retained.
```
This is not true. The `IEnumerable` generated by the compiler does not execute anything until the first `MoveNext` call (which is why you have to do the null checks in the wrapper functions instead of the Impl one).

fsateler commented on MoreLinq/ExceptBy.cs:
```diff
@@ -150,11 +154,10 @@ static partial class MoreEnumerable {
-                keys.Add(key);
+                if (!keys.Contains(key)) {
+                    yield return element;
+                    keys.Add(key);
```
[Add returns a boolean indicating whether the element was already present](https://msdn.microsoft.com/en-us/library/bb353005(v=vs.110).aspx) So no need to invoke both Contains and Add

fsateler commented on MoreLinq/IntersectAll.cs:
```diff
@@ -0,0 +1,182 @@
+        /// <returns>Sequence of elements from first sequence
+        /// whose keys are in second sequence, according to the given key selector.</returns>
+        /// <remarks>This is a sequence operation; if multiple elements in <paramref name="first"/> have
+        /// equal keys, all such elements are returned.
```
What about `second` duplicates? I think it should be mentioned that duplicates there are discarded.

fsateler commented on MoreLinq/IntersectBy.cs:
```diff
@@ -146,10 +146,10 @@ partial class MoreEnumerable {
                 if (keys.Contains(k)) {
-                    yield return item;
+                    yield return element;
                     keys.Remove(k);
```
[HashSet.Remove also reports success](https://msdn.microsoft.com/en-us/library/bb342694(v=vs.110).aspx). I know, not in your patch, but if you are already modifying...

fsateler commented:
@JamesFaix I think it is better if you split the changes into separate pull requests, as it makes reviewing the changes easier.

BTW, there are a few member names with lowercase names, the convention is using CapitalizedFirst.

Also, please check the xmldoc comments. I left a few comments on a couple, but they apply to most.

JamesFaix commented:
Thanks for all the feedback.  Sorry about the pull request bloat, I'm pretty new to Git.

I want to make sure I understand all your comments correctly.  Are you suggesting:
0. Generally, review XML docs for inconsistencies and missing members.
1. Add a very explicit `remark` to all the "All" methods, so they don't get confused with the standard set operators.
2. Add explicit comments about `second` duplicates being discarded on "All" methods.
3. Simplify `ExceptByImpl`, since `HashSet.Add` returns a `bool` that can be used instead of `Contains`.

I am a little confused about this one:
```
/// This operator uses deferred execution and streams the results, although
/// a set of keys from <paramref name="second"/> is immediately selected and retained.
```

_This is not true. The IEnumerable generated by the compiler does not execute anything until the first MoveNext call (which is why you have to do the null checks in the wrapper functions instead of the Impl one)._

Are you saying that the `HashSet` storing `second` is also deferred?

How is that related to placing null checks?

fsateler commented:
0. Correct, and for using references where appropriate.
1. I would find it useful, yes (a naive user might wonder why it exists at all). I think a simple "Like `Except`, but preserves distinct elements in the first sequence" or something like that should make it clear enough.
2. Yes, right after you document that duplicate elements in first are preserved.
3. Yes.

> Are you saying that the `HashSet` storing `second` is also deferred?

Yes, it will be executed upon fetching the first element of the resulting sequence. That is:

```
var q = a.ExceptAll(b);
// b has not been iterated over yet
q.GetEnumerator().MoveNext(); // b is iterated fully here
```

> How is that related to placing null checks?

The null checks are placed in the wrapper functions because you want the exceptions to occur when you create the sequence, not when you iterate over it, which might be in a completely separate module. If they are placed in the Impl methods, then you would get the exception upon enumerating, not upon function call. You can test it on LINQPad:

```
IEnumerable<object> Foo(int a) {
    if (a == 0) throw new Exception("Exception!");
    yield return null;
}

var q = Foo(0); // no exception here
q.Dump(); // Exception here
```

JamesFaix commented:
Got it. Thanks.  I was putting the null checks on the wrapper methods to minimize the number of checks each call had to perform; I hadn't thought about the execution timing.

JamesFaix commented:
1. Reverted all but my first 3 commits, which all deal with Except and Intersect variations.  
2. Changed the implementation of ExceptBy as mentioned above. 
3. Reviewed and corrected docs for all the Except and Intersect methods.

fsateler commented:
@JamesFaix I suggest you do the following:

1. Create a new (local) branch, where you keep all your new changes (ie, those unrelated to this PR).
2. Then do `git rebase -i master` on this branch. On the list git will present to you, just delete the lines referring to commits that are unrelated to this PR (be sure to also delete all the revert commits). After closing then git will leave you at a branch with only those changes.
3. Check it all worked out ok
4. `git push --force`. This will update this PR to have only the important commits, so it will have a few, not 20 :)

Yes, git can be confusing at first ;). Obligatory XKCD reference: http://xkcd.com/1597/

JamesFaix commented:
@fsateler Thanks for the tip.  That comic summarizes the impression I've been gathering from YouTube and StackOverflow.

fsateler commented:
@JamesFaix Why did you close the PR? I hope it wasn't due to me being annoying...

fsateler commented:
heh, I just saw v3 at #169 . Sorry for the noise.

JamesFaix commented:
No worries.  You've been very helpful, I'm just not fluent in git shell yet.
